### PR TITLE
Hide modding-related types from TypeHelper.FindType

### LINKED ISSUE #64: NML "tell" via loaded types
LogiX can be used to make a user write a type such as `NeosModLoader.NeosMod`, `NeosModSettings.NeosModSettings`, etc into a TypeField and checking if it succeeded. This can be used to detect if users are running NML.

### PR DESCRIPTION
**Pull request by zkxs**
This fixes #64.

The core idea in this PR is right as NML is spinning up, I record all currently loaded assemblies, which I call `initialAssemblies`. Harmony may or may not be loaded yet depending on if the user has it in nml_mods or not. So `initialAssemblies` - NML - Harmony is `neosAssemblies`. Any assembly not in `neosAssemblies` will be hidden from Neos's main string-to-Type conversion function, `TypeHelper.FindType()`.

As an additional safety check, I calculate assemblies that belong to mods as well. If a type belongs to neither `neosAssemblies` or `modAssemblies`, that's an indication that a type was loaded *very* late (after NML) but it's being used in game. That's pretty weird, so I have a warn log for it.

### CODE REVIEW

**Comment by EIA485** [comment]
there should be a nml config option to disable this feature, maybe "unsafe_types"

**Comment by zkxs** [comment]
> there should be a nml config option to disable this feature, maybe "unsafe_types"

Implemented now as a `hidemodtypes=true` config.

**Comment by EIA485** [comment]
ideally this harmony patch would not run in LocalWorlds, namely userspace.
there is no risk of other users running logix in on you in a local world and this patch could break some mods. 
say a mod author defined a type config, then a user is using the mod settings mod. the user can not set the type to one that is not built into neos.

File: NeosModLoader/AssemblyHider.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+		private static void FindTypePostfix(ref Type? __result)
+		{
+			if (__result != null && !neosAssemblies!.Contains(__result.Assembly))
+			{
+				if (!modAssemblies!.Contains(__result.Assembly))
+				{
+					// an assembly was in neither neosAssemblies nor modAssemblies
+					// this implies someone late-loaded an assembly after NML, and it was later used in-game
+					// this is super weird, and probably shouldn't ever happen... but if it does, I want to know about it.
```
Comment (EIA485) [inline NeosModLoader/AssemblyHider.cs:58]: wont any plugin loaded after nml is loaded cause this?

File: NeosModLoader/AssemblyHider.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+				MethodInfo target = AccessTools.DeclaredMethod(typeof(TypeHelper), nameof(TypeHelper.FindType));
+				MethodInfo patch = AccessTools.DeclaredMethod(typeof(AssemblyHider), nameof(FindTypePostfix));
+				harmony.Patch(target, postfix: new HarmonyMethod(patch));
+			}
+		}
+
+		private static HashSet<Assembly> GetNeosAssemblies(HashSet<Assembly> initialAssemblies)
+		{
+			initialAssemblies.Remove(Assembly.GetExecutingAssembly());
+			initialAssemblies.Remove(typeof(Harmony).Assembly);
```
Comment (EIA485) [inline NeosModLoader/AssemblyHider.cs:35]: should probably add a comment here clarifying why a specific case is needed.
in theory, if nml is the only thing a user is using that loads harmony, and the user followed the install instructions of placing harmony in nml_libs, initialAssemblies shouldn't contain harmony.
i could be wrong. just sayin its worth commenting.

**Comment by XDelta** [comment]
> Implemented now as a `hidemodtypes=true` config.
Offhand I think all the other config options have false as the default behavior of nml. I think the name should be adjusted to follow that for consistency as hiding types should be the default.

**Comment by zkxs** [comment]
> ideally this harmony patch would not run in LocalWorlds, namely userspace. there is no risk of other users running logix in on you in a local world and this patch could break some mods. say a mod author defined a type config, then a user is using the mod settings mod. the user can not set the type to one that is not built into neos.
> 
> a potential way to implement this would be to check what thread the FindType() method is being called on and compare that to a list of whitelisted localworld threads

Do we know for a fact that threads aren't in a pool that handles multiple worlds?

**Comment by zkxs** [comment]
> > Implemented now as a `hidemodtypes=true` config.
> 
> Offhand I think all the other config options have false as the default behavior of nml. I think the name should be adjusted to follow that for consistency as hiding types should be the default.

There is another config that defaults to true. [/doc/modloader_config.md](https://github.com/neos-modding-group/NeosModLoader/blob/master/doc/modloader_config.md)

The reason I named this `hidemodtypes` is because I think `showmodtypes` would be misleading, as NML either hides or takes no action.

File: NeosModLoader/AssemblyHider.cs (hunk 1 above)
Comment (zkxs) [reply NeosModLoader/AssemblyHider.cs:58]: No, plugins are loaded [before NML starts executing](https://github.com/neos-modding-group/NeosModLoader/blob/master/doc/making_mods.md#onengineinit).

File: NeosModLoader/AssemblyHider.cs (hunk 2 above)
Comment (zkxs) [reply NeosModLoader/AssemblyHider.cs:38]: Comment added in f9c6f22d. The install instructions weren't always to use nml_libs, and I don't want to punish users who followed the old steps.

**Comment by XDelta** [comment]
> > > Implemented now as a `hidemodtypes=true` config.
> > 
> > 
> > Offhand I think all the other config options have false as the default behavior of nml. I think the name should be adjusted to follow that for consistency as hiding types should be the default.
> 
> There is another config that defaults to true. [/doc/modloader_config.md](https://github.com/neos-modding-group/NeosModLoader/blob/master/doc/modloader_config.md)
> 
> The reason I named this `hidemodtypes` is because I think `showmodtypes` would be misleading, as NML either hides or takes no action.

AllowUnsafeTypeCheck, AllowUnsafeModTypeChecks or similar could work but just figured I'd mention it.

File: NeosModLoader/AssemblyHider.cs (hunk 1 above)
Comment (EIA485) [reply NeosModLoader/AssemblyHider.cs:62]: maybe we could add another config option to allow late loaded types thru our filter.

Comment (zkxs) [reply NeosModLoader/AssemblyHider.cs:62]: That's doable. I'm just not sure who'd be late loading types at all, hence the warning log.

**Review by ljoonal — APPROVED** [review]
While this seems okay, it should be noted that **fully** hiding anything in Neos from remote clients is hard to do properly. Even with this, I'm pretty sure there is still at least a possibility of a theoretical timing based detection approach, given a dedicated enough attacker.

**Pull request merged**